Zend/zend_compile.c

```diff
@@ -5529,16 +5591,54 @@ void zend_compile_prop_decl(zend_ast *ast) /* {{{ */
 
 		if (value_ast) {
 			zend_const_expr_to_zval(&value_zv, value_ast);
-		} else {
+
+			if (optional_type && !Z_CONSTANT(value_zv)) {
+			    if (allow_null && Z_TYPE(value_zv) == IS_NULL) {
+					/* pass */
+				} else if (optional_type == IS_ARRAY) {
+					if (Z_TYPE(value_zv) != IS_ARRAY) {
+						zend_error_noreturn(E_COMPILE_ERROR,
+							"Default value for properties with array type can only be an array");
+					}
+				} else if (optional_type == IS_CALLABLE) {
+					if (Z_TYPE(value_zv) != IS_STRING && Z_TYPE(value_zv) != IS_ARRAY) {
+						zend_error_noreturn(E_COMPILE_ERROR,
+							"Default value for properties with callable type can only be a string or an array");
```

Comment:
For parameters we only allow `null` as a `callable` default. We should be consistent here, especially as these checks do not actually ensure that the string/array is something callable.